_posts/2023-11-16-quarkus-meets-langchain4j.adoc

```diff
@@ -0,0 +1,236 @@
+
+First, you need to ingest documents into a document store. The documents are not stored as-is, but an embedding is computed. This embedding is a vector representation of the document.
+
+Then, in the Quarkus application, you need to declare the document store and the embedding to use. The extension will load the document store and the embedding and will use them to extend the LLM knowledge. Thus, before calling the LLM, it retrieves the relevant documents from the store (that's where the vector representation is useful) and attaches them to the LLM context (which essentially means adding the retrieved information from the document to the user message).
+
+The Quarkus Langchain4j extension provides facilities for both steps.
+
+The following code shows how to ingest a document into a Redis document store:
+
+[source, java]
+----
+@ApplicationScoped
+public class IngestorExample {
+
+    /**
+     * The embedding store (the database).
+     * The bean is provided by the quarkus-langchain4j-redis extension.
+     */
+    @Inject
+    RedisEmbeddingStore store;
+
+    /**
+     * The embedding model (how the vector of a document is computed).
+     * The bean is provided by the LLM (like openai) extension.
+     */
+    @Inject
+    EmbeddingModel embeddingModel;
+
+    public void ingest(List<Document> documents) {
+        EmbeddingStoreIngestor ingestor = EmbeddingStoreIngestor.builder()
+                .embeddingStore(store)
+                .embeddingModel(embeddingModel)
+                .documentSplitter(recursive(500, 0))
+                .build();  
+        ingestor.ingest(documents);
+    }
+}
+----
+
+Then, generally, in another application, you can use the populated document store to extend the LLM knowledge. First, create a bean implementing the `Retriever<TextSegment>` interface:
+
+[source, java]
+----
+@ApplicationScoped
+public class DocumentRetrieverExample implements Retriever<TextSegment> {
+
+    private final EmbeddingStoreRetriever retriever;
+
+    DocumentRetrieverExample(RedisEmbeddingStore store, EmbeddingModel model) {
+        retriever = EmbeddingStoreRetriever.from(store, model, 20);
+    }
+
+    @Override
+    public List<TextSegment> findRelevant(String s) {
+        return retriever.findRelevant(s);
+    }
+}
+----
+
+Then, add the document store and the retriever to the `@RegisterAiService` annotation:
+
+[source, java]
+----
+@RegisterAiService(retrieverSupplier = RegisterAiService.BeanRetrieverSupplier.class)
+public interface MyAiService {
+// ...
+}
+----
+
+TIP: `RegisterAiService.BeanRetrieverSupplier.class` is a special value looking for the `Retriever` bean in the Quarkus application.
+
+== Final notes
+
+This post presented the Quarkus Langchain4j extension. This is the first version of the extension, and we continue exploring what integrating LLMs into Quarkus applications means. We are looking for feedback and ideas to improve the extension. We are working removing some rough angles. 
```

Comment:
I would add some obvious but maybe not so obvious advantages this level of integration into quarkus has and what it enables:

This all fits nicely with CDI - you can basically hook up any of your CDI beans as tools or as feed into data. Camel routes can be beans, camel can use a AIService in its routes, etc.